summingbird-storm/src/main/scala/com/twitter/summingbird/storm/collector/AggregatorOutputCollector.scala

```diff
@@ -0,0 +1,121 @@
+      case (stream, _) =>
+        val tupsOut = spoutCaches(stream).tick.map { convertToSummerInputFormat(_) }
+        val tups = Await.result(tupsOut)
+        tups.foreach {
+          case (k, v) => {
+            val messageIdsTracker = streamMessageIdTracker(stream)
+            val messageIds = messageIdsTracker.remove(k)
+            val list = List(k, v).asJava.asInstanceOf[JList[AnyRef]]
+            callEmit(messageIds, list, stream)
+          }
+        }
+    }
+  }
+
+  private def convertToSummerInputFormat(flushedCache: CMap[K, V]): CMap[Int, CMap[K, V]] = {
+    flushedCache.groupBy { case (k, _) => summerShards.summerIdFor(k) }
+  }
+
+  private def emitData[K, V](cache: Future[Traversable[(Int, CMap[K, V])]], s: String): List[Int] = {
+    /*
+    The method is invoked to handle the flushed cache caused by exceeding the memoryLimit, which is called within add method.
+     */
+    var returns = MList[Int]()
+    val flushedTups = Await.result(cache)
+    flushedTups
+      .foreach {
+        case (k, v) =>
+          val messageIdsTracker = streamMessageIdTracker(s)
+          val messageIds = messageIdsTracker.remove(k)
+          val list = List(k, v).asJava.asInstanceOf[JList[AnyRef]]
+          val taskIds = callEmit(messageIds, list, s)
+          if (taskIds != null) returns ++= taskIds.asInstanceOf[JList[Int]].asScala.toList
+      }
+    returns.toList
+  }
+
+  private def callEmit(messageIds: Option[Any], list: JList[AnyRef], stream: String): JList[Integer] = {
+    /*
+    This is a wrapper method to call the emit with appropriate signature
+    based on the arguments.
+     */
+    (messageIds.isEmpty, stream.isEmpty) match {
+      case (true, true) => in.emit(list)
+      case (true, false) => in.emit(stream, list)
+      case (false, true) => in.emit(list, messageIds)
+      case (false, false) => in.emit(stream, list, messageIds)
+    }
+  }
+
+  private def add(tuple: (K, V), s: String, o: Option[Any] = None) = {
+    if (o.isDefined)
+      trackMessageId(tuple, o.get, s)
+    addToCache(tuple, s)
+  }
+
+  private def addToCache(tuple: (K, V), s: String) = {
+    val cache = spoutCaches.get(s)
+    cache match {
+      case Some(cac) => cac.add(tuple)
+      case None => {
+        spoutCaches(s) = summerBuilder.getSummer[K, V](implicitly[Semigroup[V]])
+        spoutCaches(s).add(tuple)
+      }
+    }
+  }
+
+  private def trackMessageId(tuple: (K, V), o: scala.Any, s: String): Unit = {
+    val messageIdTracker = streamMessageIdTracker.getOrElse(s, MMap[Int, MList[Object]]())
+    var messageIds = messageIdTracker.getOrElse(summerShards.summerIdFor(tuple._1), MList())
+    messageIdTracker(summerShards.summerIdFor(tuple._1)) = (messageIds.+=(o.asInstanceOf[Object]))
+    streamMessageIdTracker(s) = messageIdTracker
+  }
+
+  def extractAndProcessElements(s: String, list: JList[AnyRef], o: Option[Any] = None): JList[Integer] = {
+
+    val first: K = transform(list).get(0).asInstanceOf[K]
```

Comment:
Do we need to call transform(list) twice? Can we do it once and use it at both places?